app/models/user.rb

```diff
@@ -168,6 +171,29 @@ def change_password(oldpwd, newpwd)
     end
   end
 
+  def locked?
+    ::Settings.authentication.max_failed_login_attempts.positive? && failed_login_attempts >= ::Settings.authentication.max_failed_login_attempts
+  end
+
+  def unlock!
+    update(:failed_login_attempts => 0)
+  end
+
+  def fail_login!
+    update(:failed_login_attempts => failed_login_attempts + 1)
+
+    if locked?
+      queue_opts = {
+        :class_name  => self.class.name,
+        :instance_id => id,
+        :method_name => 'unlock!',
+        :deliver_on  => Time.now.utc + ::Settings.authentication.locked_account_timeout
+      }
+
+      MiqQueue.put_or_update(queue_opts)
```

Comment:
I'd rather we didn't use `put_or_update` as we are trying to get rid of these.  Can we just leave these unlock values on the queue separate?  I think the way the logic is set up, this will only get one on the queue anyway, since we don't call fail_login! again once the user is already locked.

Reply:
True, so should I just use:
```suggestion
      MiqQueue.put(queue_opts)
```
?

Reply:
Yes.  Once this is in, I think this is good to go.